src/Orleans/Messaging/IncomingMessageBuffer.cs

```diff
@@ -1,44 +1,329 @@
+        }
+
+        private int CalculateKnownMessageSize()
+        {
+            return headerLength + bodyLength;
+        }
+
+
+        protected List<ArraySegment<byte>> DuplicateBuffer(List<ArraySegment<byte>> body)
+        {
+            var dupBody = new List<ArraySegment<byte>>(body.Count);
+            foreach (ArraySegment<byte> seg in body)
+            {
+                var dupSeg = new ArraySegment<byte>(BufferPool.GlobalPool.GetBuffer(), seg.Offset, seg.Count);
+                Buffer.BlockCopy(seg.Array, seg.Offset, dupSeg.Array, dupSeg.Offset, seg.Count);
+                dupBody.Add(dupSeg);
+            }
+
+            return dupBody;
+        }
+
+        private void LogMessageSize(Message msg)
+        {
+            if (headerLength + bodyLength > Message.LargeMessageSizeThreshold)
+            {
+                Log.Info(ErrorCode.Messaging_LargeMsg_Incoming,
+                    "Receiving large message Size={0} HeaderLength={1} BodyLength={2}. Msg={3}",
+                    headerLength + bodyLength, headerLength, bodyLength, msg.ToString());
+                if (Log.IsVerbose3) Log.Verbose3("Received large message {0}", msg.ToLongString());
+            }
+        }
+
+        // we only care about ability to read headers length, if message body couldn't be fitted in - 
+        // the prefix will be handled on the consequent call
+        private bool TryHandleMessageLengthPrefix()
+        {
+            return TryHandlePrefix(Message.LENGTH_HEADER_SIZE);
+        }
+
+        private bool TryHandleMessageBodyPrefix()
+        {
+            return TryHandlePrefix(CalculateKnownMessageSize());
+        }
+
+        // If only part of message has arrived - store it and return true.
+        private bool TryHandlePrefix(int readLength)
+        {
+            if (AvailableReadLength < readLength)
+            {
+                prefixHolder.HandlePrefix(readBuffer, decodeOffset, receiveOffset - decodeOffset);
+                return true;
+            }
+
+            return false;
+        }
+
+        // Composes the result from stored (if any) prefix and recieved buffer.
+        private List<ArraySegment<byte>> BuildSegmentListWithLengthLimit(int length)
+        {
+            if (!prefixHolder.HasPrefix)
+            {
+                // just read from socket buffer directly
+                var res = ByteArrayBuilder.BuildSegmentListWithLengthLimit(readBuffer, decodeOffset, length);
+                decodeOffset += length;
+                return res;
+            }
+
+            int bytesRead;
+
+            // Read from the prefix holder
+            var result = new List<ArraySegment<byte>>(readBuffer.Count);
+            ByteArrayBuilder.BuildSegmentListWithLengthLimit(
+                result,
+                prefixHolder.GetPrefix(),
+                decodeOffset,
+                length,
+                out bytesRead);
+
+            length -= bytesRead;
+            decodeOffset += bytesRead;
```

Comment:
Decode offset now does not make a difference between stored and network buffers. I was thinking about hiding the buffer implementation details in the separate buffer, but this was the most obvious solution.